model/riscv_vmem_utils.sail

```diff
@@ -0,0 +1,261 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+/* This file implements utility functions for accessing memory that
+ * can be used by instruction definitions.
+ *
+ * These memory accessors performs address translation and return a
+ * Retire_Failure on error.  In order to support RVWMO memory model,
+ * address and when possible data arguments are passed as their source
+ * registers.  The `vmem_read` read accessor uses Read(Data) accesses,
+ * while the `vmem_write_*` variants use Write(Data) accesses.
+ *
+ * All addresses are provided in integer source register arguments.
+ * For write accessors, there are functions for each kind of source
+ * data register type (integer, floating point, vector).
+ *
+ * The external API of this module is:
+ *
+ *   check_misaligned - checks for the alignment of a virtual address given platform settings
+ *
+ *   vmem_read        - reads data of the specified width
+ *   vmem_write       - writes a specified data payload
+ */
+
+/* This is a external option that controls the order in which the model
+ * performs misaligned accesses.
+ */
+function sys_misaligned_order_decreasing() -> bool = config memory.misaligned.order_decreasing
+
+/* This is an external option that, when true, causes all misaligned accesses
+ * to be split into single byte operations.
+ */
+function sys_misaligned_to_byte() -> bool = config memory.misaligned.to_byte
+
+/* This is an external option that returns an integer N, such that
+ * when N is greater than zero, misaligned accesses to physical memory
+ * (as atomic events) are allowed provided the access occurs within a N
+ * byte region. N must be a power of two.
+ *
+ * This option cannot be unlimited (i.e. allowing all misaligned accesses
+ * to be atomic) as this would be incorrect above the maximum supported
+ * page size.
+ */
+function sys_misaligned_allowed_within() -> nat = config memory.misaligned.allowed_within
+
+/* Check if an 'n byte access for an address is within an aligned 'r byte region */
+val access_within : forall 'width 'n 'r, 1 <= 'n <= 'r. (bits('width), int('n), int('r)) -> bool
+
+function access_within(addr, bytes, region_bytes) = {
+  let low = addr & ~(get_slice_int(length(addr), region_bytes - 1, 0));
```

Comment:
I think `to_bits()` is exactly what this `get_slice_int` is doing?

Reply:
This is not needed after switching to the exponent-based version.